File: Single-Number-II/Single-Number-II.py

class Solution:
    # @param A, a list of integer
    # @return an integer
    def singleNumber(self, A):
        ones=0
        twos=0
        threes=0
        for i in range(len(A)):
            threes=twos & A[i]
            twos = twos | ones & A[i]
            ones = ones | A[i]
            twos= twos & ~threes
            ones =ones & ~threes
        return ones
 

File: solutions/Single-Number-II/Single-Number-II.cpp

class Solution {
public:
    int singleNumber(int A[], int n) {
        int ones=0;
        int twos=0;
        int threes=0;
        for(int i=0;i<n;i++)
        {
            twos=twos | ones&A[i];//出现两次1
            ones=ones ^ A[i];//出现一次1
            threes=ones & twos;//出现三次1
            ones = ones & ~threes;//从ones中将出现三次的位置清零
            twos = twos & ~threes;//从twos中将出现三次的位置清零
        }
        return ones;
    }
};



 